pkgs/dart_mcp/lib/src/api/api.dart

```diff
@@ -191,3 +188,107 @@ extension type PaginatedResult._fromMap(Map<String, Object?> _value)
     implements Result {
   Cursor? get cursor => _value['cursor'] as Cursor?;
 }
+
+/// Could be either [TextContent], [ImageContent] or [EmbeddedResource].
+///
+/// Use [isText], [isImage] and [isEmbeddedResource] before casting to the more
+/// specific types, or switch on the [type] and then cast.
+///
+/// Doing `is` checks does not work because these are just extension types, they
+/// all have the same runtime type (`Map<String, Object?>`).
+extension type Content._(Map<String, Object?> _value) {
+  factory Content.fromMap(Map<String, Object?> value) {
+    assert(value.containsKey('type'));
+    return Content._(value);
+  }
+
+  /// Whether or not this is a [TextContent].
+  bool get isText => _value['type'] == TextContent.expectedType;
+
+  /// Whether or not this is a [ImageContent].
+  bool get isImage => _value['type'] == ImageContent.expectedType;
+
+  /// Whether or not this is an [EmbeddedResource].
+  bool get isEmbeddedResource =>
+      _value['type'] == EmbeddedResource.expectedType;
+
+  /// The type of content.
+  ///
+  /// You can use this in a switch to handle the various types (see the static
+  /// `expectedType` getters), or you can use [isText], [isImage], and
+  /// [isEmbeddedResource] to determine the type and then do the cast.
+  String get type => _value['type'] as String;
+}
+
+/// Text provided to an LLM.
+///
+// TODO: implement `Annotated`.
+extension type TextContent.fromMap(Map<String, Object?> _value)
+    implements Content {
+  static const expectedType = 'text';
+
+  factory TextContent({required String text}) =>
+      TextContent.fromMap({'text': text, 'type': expectedType});
+
+  String get type {
+    final type = _value['type'] as String;
+    assert(type == expectedType);
+    return type;
+  }
+
+  /// The text content.
+  String get text => _value['text'] as String;
+}
+
+/// An image provided to an LLM.
+///
+// TODO: implement `Annotated`.
```

Comment:
should this todo be with the commented out annotated code in resources.dart?

Reply:
I think I should move Annotated into this file actually.